### PR TITLE
feat: add github actions and development doc

### PR DESCRIPTION
**Pull request by SerenaTiede-Zen**
github actions will run tests and linting on the code. Snyk should be setup to run on pull requests.

### CODE REVIEW

**Review by grosser — COMMENTED** [review]
can we do a metric build with `make ${{matrix.task}}` ?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      - main
+  push:
+    branches:
+      - main
+jobs:
+  test:
+    strategy:
+      matrix:
+        go-version: [ 1.17.x ]
+    runs-on: ubuntu-latest
```
Comment (notchairmk) [inline .github/workflows/main.yml]: do we need to specify zendesk runners?

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+upstream        git@github.com:West-Coast-Devops/scheduled-scaler.git (fetch)
+upstream        git@github.com:West-Coast-Devops/scheduled-scaler.git (push)
+```
+
+## Contributing upstream
+
+Run the following commands to have git setup.
+
+1. `git remote add upstream git@github.com:West-Coast-Devops/scheduled-scaler.git`
+2. `git checkout -b ${your-branch-name-here} upstream/master`
```
Comment (notchairmk) [inline DEVELOPMENT.md]: nit: maybe `git fetch upstream` before checkout?

File: .github/workflows/main.yml (hunk 1 above)
Comment (SerenaTiede-Zen) [reply .github/workflows/main.yml]: since this is a public repo we can use the normal runners

Comment (notchairmk) [reply .github/workflows/main.yml]: 🤦

File: pkg/services/scaling/metadata/cluster.go [hunk 3]
```diff
@@ -15,8 +15,18 @@ func GetClusterInfo() (projectId, zone string, err error) {
...
+	}()
+	if err != nil {
+		return "", "", err
+	}
+	defer func() {
+		err = projectIdResp.Body.Close()
+	}()
+	if err != nil {
+		return "", "", err
+	}
```
Comment (jklaw90) [inline pkg/services/scaling/metadata/cluster.go]: the err check here happens before the defer statement so it'd be checking w/e happened before this?

i think we could just defer the close on both these without error checking?
do we really care if we processed everything?

File: pkg/services/scaling/metadata/cluster.go [hunk 4]
```diff
@@ -15,8 +15,18 @@ func GetClusterInfo() (projectId, zone string, err error) {
...
 	projectIdBody, err := ioutil.ReadAll(projectIdResp.Body)
```
Comment (jklaw90) [inline pkg/services/scaling/metadata/cluster.go]: should handle error here

File: pkg/services/scaling/metadata/cluster.go [hunk 5]
```diff
@@ -15,8 +15,18 @@ func GetClusterInfo() (projectId, zone string, err error) {
...
 	projectId = string(projectIdBody)
 	zoneBody, err := ioutil.ReadAll(zoneResp.Body)
```
Comment (jklaw90) [inline pkg/services/scaling/metadata/cluster.go]: should handle error here. this would panic on `zone = zoneSlice[len(zoneSlice)-1]`

File: pkg/client/clientset/versioned/scheme/register.go [hunk 6]
```diff
@@ -48,5 +48,8 @@ func init() {
 // After this, RawExtensions in Kubernetes types will serialize kube-aggregator types
 // correctly.
 func AddToScheme(scheme *runtime.Scheme) {
-	scalingv1alpha1.AddToScheme(scheme)
+	err := scalingv1alpha1.AddToScheme(scheme)
```
Comment (notchairmk) [inline pkg/client/clientset/versioned/scheme/register.go]: would it be preferable to panic during init if returning `err`?

File: pkg/client/clientset/versioned/scheme/register.go [hunk 7]
```diff
@@ -48,5 +48,8 @@ func init() {
...
+	if err != nil {
```
Comment (jklaw90) [inline pkg/client/clientset/versioned/scheme/register.go]: should we do something with the error? if its not going to happen `// nolint:errcheck`

File: pkg/signals/signal_posix.go [hunk 8]
```diff
@@ -1,3 +1,4 @@
+//go:build !windows
```
Comment (jklaw90) [inline pkg/signals/signal_posix.go]: nice

File: .github/workflows/main.yml (hunk 1 above)
Comment (a7i) [reply .github/workflows/main.yml]: But we're using GitHub Actions such as `zendesk/setup-go@v2` already

File: pkg/services/scaling/metadata/cluster.go (hunk 3 above)
Comment (SerenaTiede-Zen) [reply pkg/services/scaling/metadata/cluster.go]: one of the linters requires all errors to be checked which leads to weird syntax. I could disable the err check linter for this bit

Comment (jklaw90) [reply pkg/services/scaling/metadata/cluster.go]: i think we can use `// nolint:errcheck`

**Comment by SerenaTiede-Zen** [comment]
> can we do a metric build with `make ${{matrix.task}}` ?

done

File: pkg/client/informers/externalversions/generic.go [hunk 9]
```diff
@@ -51,11 +51,9 @@ func (f *genericInformer) Lister() cache.GenericLister {
 // ForResource gives generic access to a shared informer of the matching type
 // TODO extend this to unknown resources with a client pool
 func (f *sharedInformerFactory) ForResource(resource schema.GroupVersionResource) (GenericInformer, error) {
-	switch resource {
 	// Group=scaling.k8s.restdev.com, Version=v1alpha1
-	case v1alpha1.SchemeGroupVersion.WithResource("scheduledscalers"):
+	if v1alpha1.SchemeGroupVersion.WithResource("scheduledscalers") == resource {
 		return &genericInformer{resource: resource.GroupResource(), informer: f.Scaling().V1alpha1().ScheduledScalers().Informer()}, nil
```
Comment (grosser) [inline pkg/client/informers/externalversions/generic.go]: prefer to leave flavor things like this alone to reduce diff with upstream, or make a separate PR in upstream to get it merged first

File: pkg/services/scaling/metadata/cluster.go (hunk 3 above)
Comment (jklaw90) [reply pkg/services/scaling/metadata/cluster.go]:
```suggestion
	defer zoneResp.Body.Close() // nolint:errcheck
	defer projectIdResp.Body.Close() // nolint:errcheck
```

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+          path: /home/runner/.cache/go-build
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: golangci-lint
+        uses: zendesk/golangci-lint-action@v2
+        with:
+          version: v1.42
+      - name: go ${{ matrix.go-version }} target ${{ matrix.task }}
+        run: OPERATOR=scaling make ${{ matrix.task }}
```
Comment (grosser) [inline .github/workflows/main.yml:35]: can OPERATOR be set by default ? ... are there other valid values for it ?

Comment (SerenaTiede-Zen) [reply .github/workflows/main.yml:35]: we could set it up that way but that's how the upstream had theirs setup

**Review by grosser — APPROVED** [review]
- this should all go upstream (except for DEVELOPMENT.md)
- prefer using a fork for behavior changes so the diff and rebase is simpler (can still make upstream PRs, just don't use them until they are merged)

File: pkg/client/clientset/versioned/scheme/register.go (hunk 6 above)
Comment (SerenaTiede-Zen) [reply pkg/client/clientset/versioned/scheme/register.go]: I think so but I think I'll revert some of these changes to make the linter happy, and move those into a separate pr so i can yeet things upstream

**Comment by SerenaTiede-Zen** [comment]
heads up I'm going to merge knowing that the lint is going to fail
